### ISSUE TITLE
Job list Duration field is sorted incorrectly

### ISSUE THREAD

**Issue by kingsd041** [rancher/dashboard#6899]
**Setup**
- Rancher version: v2.6.8
- Browser type & version: chrome 105.0.5195.125

**Describe the bug**

Job list Duration field is sorted incorrectly

**To Reproduce**

1. Navigate to the Job page
2. Click Duration to sort

**Result**
Not sorted by time

**Screenshots**

<img width="1786" alt="image" src="https://user-images.githubusercontent.com/19684717/190360392-8390b678-fb49-4ceb-8a8f-98f39d1df02c.png">

<img width="1789" alt="image" src="https://user-images.githubusercontent.com/19684717/190360453-a6134310-cefe-4afa-9bde-b7fabfeb6a3c.png">

**Others**

Same problem with Last Schedule field in CronJobs

**Comment by brudnak** [comment]
# ✅ PASSED

## Reproduction Environment




| Component                        | Version / Type                          |
| -------------------------------- | --------------------------------------- |
| Rancher version                  | 2.7.0                                   |
| Installation option              | Helm (high availability)                |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                            |
| Cert Details                     | Let's Encrypt / `traefik`               |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2    |
| Helm version                     | v2.16.8-rancher2                        |
| Downstream cluster type          | not applicable                          |
| Downstream K8s version           | not applicable                          |
| Authentication providers enabled | local                                   |
| Logged in user role              | Administrator                           |
| Browser type                     | Google Chrome                           |
| Browser version                  | 108.0.5359.98 (Official Build) (x86_64) |

<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>


## Reproduction steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click the `hamburger menu` >>> `local` >>> `Workload` >>> `Jobs` >>> filter to `default` namespace
3. Click the `Kubectl Shell` button and wait for it to connect
4. Once connected run `vim job.yml`
5. Paste in the following yaml and save

>Here is an example Job config. It computes π to 2000 places and prints it out. It takes around 10s to complete.


```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

6. Run this a few times, changing the name `test1` >>> `test2` >>> `test3` etc each time
7. Also make this number larger over time `print bpi(2000)` to get longer durations
8. Click the `Duration` column so that it's `▲`
9. Noticed the durations were out of order like the following:

```txt
9s
9s
9s
9s
24s
14s
```



## Additional Info

## RESULTS

#### ✅ Expected
Expected the duration column to be sorted correctly

#### ❌ Actual
The duration column was not sorted correctly

<hr>

## Validation Environment



| Component                        | Version / Type                                                                     |
| -------------------------------- | ---------------------------------------------------------------------------------- |
| Rancher version                  | v2.7-216bb7f5fc36e93ef5df2a47aaf10ce592e697f2-head                                 |
| Rancher commit link              | https://github.com/rancher/rancher/commit/216bb7f5fc36e93ef5df2a47aaf10ce592e697f2 |
| Installation option              | Helm (high availability)                                                           |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                                                                       |
| Cert Details                     | Let's Encrypt / `traefik`                                                          |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2                                               |
| Helm version                     | v2.16.8-rancher2                                                                   |
| Downstream cluster type          | not applicable                                                                     |
| Downstream K8s version           | not applicable                                                                     |
| Authentication providers enabled | local                                                                              |
| Logged in user role              | Administrator                                                                      |
| Browser type                     | Google Chrome                                                                      |
| Browser version                  | 108.0.5359.98 (Official Build) (x86_64)                                            |


<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>

## Validation steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click the `hamburger menu` >>> `local` >>> `Workload` >>> `Jobs` >>> filter to `default` namespace
3. Click the `Kubectl Shell` button and wait for it to connect
4. Once connected run `vim job.yml`
5. Paste in the following yaml and save

>Here is an example Job config. It computes π to 2000 places and prints it out. It takes around 10s to complete.


```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

6. Run this a few times, changing the name `test1` >>> `test2` >>> `test3` etc each time
7. Also make this number larger over time `print bpi(2000)` to get longer durations
8. Click the `Duration` column so that it's `▲`
9. Noticed the durations were sorted as expected ✅:

```txt
23s
10s
10s
9s
```

## Additional Info

## RESULTS

#### ✅ Expected
Expected the duration column to be sorted correctly

#### ✅ Actual
The duration column was sorted correctly


#### Additional Checks

- Tested both `▲` and `▼` sorting: PASS ✅

**Comment by brudnak** [comment]
# 🚧 REOPENED

## TLDR:

>See the last additional test case at the very bottom in the table.
>When using a script to create a bunch of k8s jobs sorting isn't working as expected.
>Screen shot at bottom as well.

## Reproduction Environment




| Component                        | Version / Type                          |
| -------------------------------- | --------------------------------------- |
| Rancher version                  | 2.7.0                                   |
| Installation option              | Helm (high availability)                |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                            |
| Cert Details                     | Let's Encrypt / `traefik`               |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2    |
| Helm version                     | v2.16.8-rancher2                        |
| Downstream cluster type          | not applicable                          |
| Downstream K8s version           | not applicable                          |
| Authentication providers enabled | local                                   |
| Logged in user role              | Administrator                           |
| Browser type                     | Google Chrome                           |
| Browser version                  | 108.0.5359.98 (Official Build) (x86_64) |

<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>


## Reproduction steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click the `hamburger menu` >>> `local` >>> `Workload` >>> `Jobs` >>> filter to `default` namespace
3. Click the `Kubectl Shell` button and wait for it to connect
4. Once connected run `vim job.yml`
5. Paste in the following yaml and save

>Here is an example Job config. It computes π to 2000 places and prints it out. It takes around 10s to complete.


```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

6. Run this a few times, changing the name `test1` >>> `test2` >>> `test3` etc each time
7. Also make this number larger over time `print bpi(2000)` to get longer durations
8. Click the `Duration` column so that it's `▲`
9. Noticed the durations were out of order like the following:

```txt
9s
9s
9s
9s
24s
14s
```



## Additional Info

## RESULTS

#### ✅ Expected
Expected the duration column to be sorted correctly

#### ❌ Actual
The duration column was not sorted correctly

<hr>

## Validation Environment



| Component                        | Version / Type                                                                     |
| -------------------------------- | ---------------------------------------------------------------------------------- |
| Rancher version                  | v2.7-216bb7f5fc36e93ef5df2a47aaf10ce592e697f2-head                                 |
| Rancher commit link              | https://github.com/rancher/rancher/commit/216bb7f5fc36e93ef5df2a47aaf10ce592e697f2 |
| Installation option              | Helm (high availability)                                                           |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                                                                       |
| Cert Details                     | Let's Encrypt / `traefik`                                                          |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2                                               |
| Helm version                     | v2.16.8-rancher2                                                                   |
| Downstream cluster type          | not applicable                                                                     |
| Downstream K8s version           | not applicable                                                                     |
| Authentication providers enabled | local                                                                              |
| Logged in user role              | Administrator                                                                      |
| Browser type                     | Google Chrome                                                                      |
| Browser version                  | 108.0.5359.98 (Official Build) (x86_64)                                            |


<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>

## Validation steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click the `hamburger menu` >>> `local` >>> `Workload` >>> `Jobs` >>> filter to `default` namespace
3. Click the `Kubectl Shell` button and wait for it to connect
4. Once connected run `vim job.yml`
5. Paste in the following yaml and save

>Here is an example Job config. It computes π to 2000 places and prints it out. It takes around 10s to complete.


```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

6. Run this a few times, changing the name `test1` >>> `test2` >>> `test3` etc each time
7. Also make this number larger over time `print bpi(2000)` to get longer durations
8. Click the `Duration` column so that it's `▲`
9. Noticed the durations were sorted as expected ✅:

```txt
23s
10s
10s
9s
```

## Additional Info

## RESULTS

#### ✅ Expected
Expected the duration column to be sorted correctly

#### ✅ Actual
The duration column was sorted correctly


#### Additional Checks

- Tested both `▲` and `▼` sorting: PASS ✅

## 📋 Additional QA Test Plan

<table>
    <tbody>
        <tr>
            <th>Scenario</th>
            <th>Steps</th>
            <th>Expected</th>
            <th>Actual & Pass/Fail</th>
        </tr>
        <tr>
            <td>Test both ▲ and ▼ on a small number of jobs with durations in the seconds range</td>
            <td>
                <ol>
                    <li>Create four jobs with similar short durations such as:</li>
                    <li>29s, 10s, 10s, 9s</li>
                    <li>Sort them both with ▲ and ▼ </li>
                </ol>
            </td>
            <td>Expected: Sorting with both ▲ and ▼ to be accurate for the durations described</td>
            <td>Actual: Sorting with both ▲ and ▼ to be accurate for the durations described<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Use a script (see script below this table) to test with a larger number of jobs and duration times</td>
            <td>
                <ol>
                    <li>Starting from the default Rancher homepage /dashboard/home</li>
                    <li>Click the hamburger menu >>> local >>> Workload >>> Jobs >>> filter to default namespace</li>
                    <li>Click the Kubectl Shell button and wait for it to connect</li>
                    <li>Create a file with vim script.sh</li>
                    <li>Paste in the script listed below this table</li>
                    <li>Save the file</li>
                    <li>Execute with the for loop bash listed below the script</li>
                    <li>For for all the k8s jobs to get created</li>
                    <li>Now filter the duration with both ▲ and ▼</li>
                </ol>
            </td>
            <td>Expected: For the sorting to be correct</td>
            <td>Actual: The sorting is not as expected durations with minutes are throwing it off. <b>see screenshot below</b><br><br>Pass/Fail: ❌<code>FAIL</code></td>
        </tr>
    </tbody>
</table>

>Script to create a bunch of k8s jobs

```shell
#!/bin/bash

# Set the new name for the job
new_name="test-name-$(date +%s)"

# Generate a random number between 2000 and 200000
new_bpi_value=$(($RANDOM % 200000 + 2000))

# Read the yaml file into a variable
yaml=$(cat <<EOM
apiVersion: batch/v1
kind: Job
metadata:
  name: test-name1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
EOM
)

# Replace the old job name and bpi value with the new ones
yaml=$(echo "$yaml" | sed "s/test-name1/$new_name/g" | sed "s/2000/$new_bpi_value/g")

# Save the modified yaml to a file
echo "$yaml" > filename.yml

# Apply the changes using kubectl
kubectl apply -f filename.yml
```

> Execute script with

```shell
for i in {1..1000}; do ./script.sh; done
```

![2022-12-12_21-09-01](https://user-images.githubusercontent.com/40877149/207226801-61c13ec8-c0a0-43a8-966f-5eab6139b926.png)

**Comment by torchiaf** [comment]
Thanks @brudnak for the huge amount of info reported during your tests!

I investigated on this issue and what I found is basically the sorting is working correctly.
The issue is that the value in the `Duration` column is not updated until the related Job is completed, but behind the scenes, the value used to calculate the sorting is correct.

As a proof of it, the running Jobs are on the bottom of the page, so the sorting is correct but the `Duration` value is still the inital one.

I just pushed another fix to make the field responsive.

**Comment by brudnak** [comment]
# ✅ PASSED

## Reproduction Environment

#### Not required.

<hr>

## Validation Environment


| Component                        | Version / Type                                                                     |
| -------------------------------- | ---------------------------------------------------------------------------------- |
| Rancher version                  | v2.7-5df9701e63e148e64c949f49fe768d56ccdfa9b5-head                                 |
| Rancher commit link              | https://github.com/rancher/rancher/commit/5df9701e63e148e64c949f49fe768d56ccdfa9b5 |
| Installation option              | Helm (high availability)                                                           |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                                                                       |
| Cert Details                     | Let's Encrypt / `traefik`                                                          |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2                                               |
| Helm version                     | v2.16.8-rancher2                                                                   |
| Downstream cluster type          | RKE1 Linode                                                                        |
| Downstream K8s version           | v1.24.8                                                                            |
| Authentication providers enabled | local                                                                              |
| Logged in user role              | Administrator, Standard                                                            |
| Browser type                     | Google Chrome                                                                      |
| Browser version                  | 108.0.5359.124 (Official Build) (x86_64)                                           |

<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>

## Validation steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click the `hamburger menu` >>> `local` >>> `Workload` >>> `Jobs` >>> filter to `default` namespace
3. Click the `Kubectl Shell` button and wait for it to connect
4. Once connected run `vim job.yml`
5. Paste in the following yaml and save

>Here is an example Job config. It computes π to 2000 places and prints it out. It takes around 10s to complete.


```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

6. Run this a few times, changing the name `test1` >>> `test2` >>> `test3` etc each time
7. Also make this number larger over time `print bpi(2000)` to get longer durations
8. Click the `Duration` column so that it's `▲`
9. Noticed the durations were sorted as expected ✅:

```txt
23s
10s
10s
9s
```

## Additional Info

## RESULTS

#### ✅ Expected
Expected the duration column to be sorted correctly

#### ✅ Actual
The duration column was sorted correctly


#### Additional Checks

- Tested both `▲` and `▼` sorting: PASS ✅

## 📋 Additional QA Test Plan

<table>
    <tbody>
        <tr>
            <th>Scenario</th>
            <th>Steps</th>
            <th>Expected</th>
            <th>Actual & Pass/Fail</th>
        </tr>
        <tr>
            <td>Test both ▲ and ▼ on a small number of jobs with durations in the seconds range</td>
            <td>
                <ol>
                    <li>Create four jobs with similar short durations such as:</li>
                    <li>29s, 10s, 10s, 9s</li>
                    <li>Sort them both with ▲ and ▼ </li>
                </ol>
            </td>
            <td>Expected: Sorting with both ▲ and ▼ to be accurate for the durations described</td>
            <td>Actual: Sorting with both ▲ and ▼ to be accurate for the durations described<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Use a script (see script below this table) to test with a larger number of jobs and duration times</td>
            <td>
                <ol>
                    <li>Starting from the default Rancher homepage /dashboard/home</li>
                    <li>Click the hamburger menu >>> local >>> Workload >>> Jobs >>> filter to default namespace</li>
                    <li>Click the Kubectl Shell button and wait for it to connect</li>
                    <li>Create a file with vim script.sh</li>
                    <li>Paste in the script listed below this table</li>
                    <li>Save the file</li>
                    <li>Execute with the for loop bash listed below the script</li>
                    <li>For for all the k8s jobs to get created</li>
                    <li>Now filter the duration with both ▲ and ▼</li>
                </ol>
            </td>
            <td>Expected: For the sorting to be correct</td>
            <td>Actual: Sorting with both ▲ and ▼ to be accurate for the durations described<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
    </tbody>
</table>

>Script to create a bunch of k8s jobs

```shell
#!/bin/bash

# Set the new name for the job
new_name="test-name-$(date +%s)"

# Generate a random number between 2000 and 200000
new_bpi_value=$(($RANDOM % 200000 + 2000))

# Read the yaml file into a variable
yaml=$(cat <<EOM
apiVersion: batch/v1
kind: Job
metadata:
  name: test-name1
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl:5.34.0
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
EOM
)

# Replace the old job name and bpi value with the new ones
yaml=$(echo "$yaml" | sed "s/test-name1/$new_name/g" | sed "s/2000/$new_bpi_value/g")

# Save the modified yaml to a file
echo "$yaml" > filename.yml

# Apply the changes using kubectl
kubectl apply -f filename.yml
```

> Execute script with

```shell
for i in {1..1000}; do ./script.sh; done
```
